### PR TITLE
feat: add terraform ci

### PR DESCRIPTION
**Pull request by AtzeDeVries**

### CODE REVIEW

File: internal/targets/terraform/tools.Dockerfile [hunk 1]
```diff
@@ -0,0 +1,3 @@
+FROM ghcr.io/terraform-linters/tflint-bundle:v0.48.0.0@sha256:523a85b4bf7af415cbb5f53ac1b56cd7501c8c8125f96c5b5cc19c895d894513 AS tflint-universal
+FROM docker.io/hashicorp/terraform:1.5.7@sha256:c3bc74e7a2a8fab8216cbbedf12a9637db09288806a6aa537b6f397cba04dd93 AS terraform-amd64
+FROM docker.io/hashicorp/terraform:1.5.7@sha256:bbc41b888559ece9b3fce0ab834ed7120d898e1a808e4716c37b2d2efb4f2782 AS terraform-arm64
```
Comment (bendiknesbo) [inline internal/targets/terraform/tools.Dockerfile:3]: These terraform-versions will be matched by Dependabot, and probably attempted to be updated. If that is fine, then I assume we could bump it to the newest already, but if it is not, maybe we should add a unit-test that verifies that the terraform-versions in this file is not newer than the maximum support version? That would prevent Dependabot-PR's from being merged.

File: internal/terraform/terraform.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/coopnorge/mage/internal/devtool"
+)
+
+
+// IsTerraformProject returns true if a directory contains a go module.
```
Comment (bendiknesbo) [inline internal/terraform/terraform.go:16]:
```suggestion
// IsTerraformProject returns true if a directory contains a Terraform project.
```

Please take a look at all the comments, there are a lot of copy-paste errors

File: internal/terraform/terraform.go [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+// Go modules.
+func FindTerraformProjects(base string) ([]string, error) {
+	directories := []string{}
+
+	err := filepath.WalkDir(base, func(workDir string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if isDotDirectory(workDir, d) {
+			return nil
```
Comment (bendiknesbo) [inline internal/terraform/terraform.go:37]: If we find a dot-directory, we should probably not go into the dot-directory to find any other terraform-projects in there. 
```suggestion
			return filepath.SkipDir
```

File: internal/terraform/terraform.go [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+	return strings.HasPrefix(path, ".")
+}
+
+
+
+// Test automates testing the packages named by the import paths, see also: go
+// test.
+func Test(directory string) error {
+	return DevtoolTerraform(nil, directory, "validate")
+}
```
Comment (bendiknesbo) [inline internal/terraform/terraform.go:69]: Once we upgrade past Terraform 1.5.7, we could also do proper tests here: https://developer.hashicorp.com/terraform/language/tests

File: internal/terraform/terraform.go [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+
+// Lint runs the linters
+func Lint(directory string) error {
+
+	err := DevtoolTerraform(nil, directory, "fmt","-diff","-check")
```
Comment (bendiknesbo) [inline internal/terraform/terraform.go:74]: NIT: Some gofmt issues here. Leading empty line, and missing spaces between arguments

File: internal/terraform/terraform.go [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+
+// Init downloads Terraform modules locally
+func Init(directory string) error {
+	log.Printf("Running terraform init for  %q", directory)
+	err := DevtoolTerraform(nil, directory, "init")
+	if err != nil {
+		return err
+	}
+	return nil
+}
```
Comment (bendiknesbo) [inline internal/terraform/terraform.go:111]: Do we need a way to add the `-upgrade` flag to `terraform init`?

File: internal/terraform/terraform.go [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+		"--volume", "$HOME/.terraform.d:/root/.terraform.d",
+		"--volume", "$HOME/.gitconfig:/root/.gitconfig",
+		"--volume", "$HOME/.ssh:/root/.ssh",
+		"--volume", fmt.Sprintf("%s:/src", cwd),
+		"--workdir", path.Join("/src",directory),
+	}
+
+	if env == nil {
+		env = map[string]string{}
+	}
```
Comment (bendiknesbo) [inline internal/terraform/terraform.go:164]: This looks redundant. It's safe to range of a `nil` map in Go.

Unless you also intend to always add a set of default env-vars?

**Review by bendiknesbo — COMMENTED** [review]
Good progress!

Some general comments:
- A lot of usages in the existing devtooling is using `terraform_init_args="-backend=false"`. Is this redundant, with Mage/Spacelift? Do we need to add support for it?
- This iteration seems to be missing some tools, like tf-docs and trivy. Do you plan to add them in this PR?

**Review by bendiknesbo — COMMENTED** [review]
Good progress!

Some general comments:
- A lot of usages in the existing devtooling is using `terraform_init_args="-backend=false"`. Is this redundant, with Mage/Spacelift? Do we need to add support for it?
- This iteration seems to be missing some tools, like tf-docs and trivy. Do you plan to add them in this PR?

File: internal/targets/terraform/tools.Dockerfile (hunk 1 above)
Comment (AtzeDeVries) [reply internal/targets/terraform/tools.Dockerfile:3]: i have been planning to see if we can add some dependabot ignore statements somehow in the dockerfile.

File: internal/terraform/terraform.go (hunk 6 above)
Comment (AtzeDeVries) [reply internal/terraform/terraform.go:111]: maybe in a separate target. Dependabot should deal with the updates, but if you require it locally we might want to expose a target for this. I think default behaviour should be to not upgrade.

File: internal/terraform/terraform.go (hunk 7 above)
Comment (AtzeDeVries) [reply internal/terraform/terraform.go:164]: hehe i just relied on Niels Hendriks knowledge here :)

**Comment by AtzeDeVries** [comment]
> Good progress!
> 
> Some general comments:
> 
>     * A lot of usages in the existing devtooling is using `terraform_init_args="-backend=false"`. Is this redundant, with Mage/Spacelift? Do we need to add support for it?

We dont define any backend anymore, this is now handled by spacelift (it automatilcy injects a backend), so it is redundant. 
> 
>     * This iteration seems to be missing some tools, like tf-docs and trivy. Do you plan to add them in this PR?

Yes they need to be added.

File: internal/devtool/devtool.go [hunk 8]
```diff
@@ -63,12 +66,34 @@ func Build(tool, dockerfile string) error {
...
+
+	switch toolAvailable := strings.Fields(scanner.Text())[len(strings.Fields(scanner.Text()))-1]; toolAvailable {
+    case archTool:
+	    return archTool, nil
+	case universalTool:
+		return universalTool, nil
+	}
+  }
+  return "", fmt.Errorf("Unable to find devtool for tool \"%s\" for the host architecture %s", tool,runtime.GOARCH)
+}
```
[the remark this answers is not in the document]
Comment (AtzeDeVries) [reply internal/devtool/devtool.go:99]: ≈

> Is this an optimization, to try to use an image from the same arch as the host?
> 
> Since our devtool-dockerfiles are minimally small, it would probably be just as fine to use `strings.Split` to get every line, instead of using `bufio.Scanner` (readability-wise).
> 
Is the main reason for not using `bufio` to not import an other library? I chose to not use string.Split because of the potential differnt ways new lines can be made and i guessed that `bufio.Scanner` would just deal with that. 


> `switch toolAvailable := strings.Fields(scanner.Text())[len(strings.Fields(scanner.Text()))-1]; toolAvailable {` could benefit from a comment stating what it is trying to extract from the line (e.g. `terraform-amd64` vs `terraform-arm64`). This also does not allow for having comments on the line. Could we use a regex instead, to make sure that we are not matching any comments?
>
Good point, that needs fixing. 

> Suggestion: Could we try to find the most accurate image, but allow a fallback, without requiring a `-universal` prefix? E.g. a dockerfile like this for terraform:
> 
> ```dockerfile
> FROM ghcr.io/terraform-linters/tflint-bundle:v0.48.0.0@sha256:523a85b4bf7af415cbb5f53ac1b56cd7501c8c8125f96c5b5cc19c895d894513 AS tflint
> FROM docker.io/hashicorp/terraform:1.5.7@sha256:c3bc74e7a2a8fab8216cbbedf12a9637db09288806a6aa537b6f397cba04dd93 AS terraform # Default Terraform-image, amd64
> FROM docker.io/hashicorp/terraform:1.5.7@sha256:bbc41b888559ece9b3fce0ab834ed7120d898e1a808e4716c37b2d2efb4f2782 AS terraform-arm64 # Optimized version of running arm64 version
> ```

I felt adding `-universal` was shows more intend. You make it deliberate for the viewer that there are universal (which should work for all archs)  images and arch specific images.

File: internal/terraform/terraform.go (hunk 3 above)
Comment (AtzeDeVries) [reply internal/terraform/terraform.go:37]: ok, good point, i did not know about SkipDir

File: internal/devtool/devtool.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (bendiknesbo) [reply internal/devtool/devtool.go:99]: >Is the main reason for not using bufio to not import an other library? I chose to not use string.Split because of the potential differnt ways new lines can be made and i guessed that bufio.Scanner would just deal with that.

I have no qualms of using `bufio` (and it's also part of the stdlib), I just thought it was conseptually easier to think about a slice of strings (1 line per string) instead of using a scanner

File: internal/targets/terraform/terraform_test.go [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+				os.RemoveAll(".terraform.lock.hcl")
+				os.RemoveAll(".terraform")
+			})
+			gotErr := Init(context.Background())
+			if tt.wantErr {
+				assert.Error(t, gotErr)
+			} else {
+				assert.NoError(t, gotErr)
+			}
+		})
```
Comment (bendiknesbo) [inline internal/targets/terraform/terraform_test.go:42]: This is failing due to a behaviour in Mage. I have reported it upstream: https://github.com/magefile/mage/issues/524

Comment (AtzeDeVries) [reply internal/targets/terraform/terraform_test.go]: thanks for figuring this out, was a real hard blocker.

File: internal/targets/terraform/tools.Dockerfile [hunk 10]
```diff
@@ -0,0 +1,3 @@
+FROM ghcr.io/terraform-linters/tflint:v0.57.0 AS tflint
+FROM docker.io/hashicorp/terraform:1.5.7@sha256:9fc0d70fb0f858b0af1fadfcf8b7510b1b61e8b35e7a4bb9ff39f7f6568c321d AS terraform
+FROM docker.io/aquasec/trivy:0.62.0@sha256:cb84170aa1fb6942c11b638969dce0845735c35f40d3bd48851c7f2d83a3c1ae AS trivy
```
Comment (bendiknesbo) [inline internal/targets/terraform/tools.Dockerfile:3]: Trivy 0.62.0 has a known bug that affected us

```suggestion
FROM docker.io/aquasec/trivy:0.62.1@sha256:fc10faf341a1d8fa8256c5ff1a6662ef74dd38b65034c8ce42346cf958a02d5d AS trivy
```

File: 2
[the remark this answers is not in the document]
Comment (AtzeDeVries) [reply 2:1]: no that is intended, fro now we dont want terraform above 1.6 because of licensing issues.

[the remark this answers is not in the document]
Comment (bendiknesbo) [reply 2:1]: No, this comment is for the file with the name `2`, not the dependabot-change

[the remark this answers is not in the document]
Comment (AtzeDeVries) [reply 2:1]: ah, missread that. will fix

[the remark this answers is not in the document]
Comment (AtzeDeVries) [reply 2:1]: https://github.com/coopnorge/mage/pull/155/commits/065b8df3cfd1900def3d855e1f19bade89b84642
